src/main/scala/is/hail/utils/richUtils/RichRDD.scala

```diff
@@ -99,4 +97,40 @@ class RichRDD[T](val r: RDD[T]) extends AnyVal {
         r.compute(split.asInstanceOf[SubsetRDDPartition].parentPartition, context)
     }
   }
-}
+
+  def headPerPartition(n: Int)(implicit ct: ClassTag[T]): RDD[T] = {
+    require(n >= 0)
+    r.mapPartitions({ it => it.take(n) }, preservesPartitioning = true)
+  }
+
+  def head(n: Long)(implicit ct: ClassTag[T]): RDD[T] = {
+    require(n >= 0)
+
+    val sc = r.sparkContext
+    val parallelism = sc.defaultParallelism
+    val nPartitions = r.partitions.length
+
+    val counts = new Array[Long](nPartitions)
+    var nScanned = 0
+    var nSeen = 0L
+
+    while (nSeen < n && nScanned < nPartitions) {
+      val scanPartitions = nScanned until (nScanned + parallelism): Seq[Int]
```

Comment:
A lot of times, parallelism is actually lower than the real number of cores. This is definitely true using Yarn as a scheduler, which both Google Dataproc and our on-prem cluster do. I think the number scanned should grow geometrically, not arithmetically - 4x growth per iteration is what Spark does, that seems reasonable.